### ISSUE TITLE
Lookahead bias in `compute_rsi()`

### ISSUE THREAD

**Issue by miohtama** [FranQuant/AI-based-Trading-Strategies#10]
Currently, in the notebook `04_ATT_LSTM_model_template`, the function `compute_rsi` has a forward-looking bias.

- I feel 75% confident I nailed the cause correctly here, I could be still missing something in my analysis
- The function is incorporating future data in the indicator calculations
- This causes the backtest to access data that should be available yet at the time of deciding a trade
- This causes the backtest results to be too profitable
- Below is a fixed function `compute_rsi()` function

## How to test for forward-looking bias

- Calculate the technical indicators for all indicators for your price dataframe
- Now cut the price dataframe at an arbitrary date halfway ("cut off date")
- Calculate the technical indicators for the cut data
- The indicator values should be the same for the cut-off data in both uncut and cut data: 
      1. uncut data at the cut-off date and...
      2. the last row of cut data
-> If the uncut data differs, it means some of the indicators used future values in their calculations

## Fixed function

This one should not have forward-looking bias:

```python
    def compute_rsi(df, column="adjusted_close", period=14):
        """RSI calculation using simple moving average."""
        prices = df[column]
        delta = prices.diff()
        gains = delta.where(delta > 0, 0)
        losses = -delta.where(delta < 0, 0)

        # Use simple moving average for the first calculation
        avg_gains = gains.rolling(window=period).mean()
        avg_losses = losses.rolling(window=period).mean()

        rs = avg_gains / avg_losses
        rsi = 100 - (100 / (1 + rs))

```

## Explanation of bias

(Mostly copy pasted from Grok)

```
    # Momentum Indicators: RSI & MACD
    def xxx_lookahead_bias_compute_rsi(self, df, column="adjusted_close", period=14):
        delta = df[column].diff()
        gain = np.where(delta > 0, delta, 0)
        loss = np.where(delta < 0, -delta, 0)
        avg_gain = np.convolve(gain, np.ones(period)/period, mode='same')
        avg_loss = np.convolve(loss, np.ones(period)/period, mode='same')
        rs = avg_gain / avg_loss
        return 100 - (100 / (1 + rs))
```

Forward-looking bias in the provided function, `xxx_lookahead_bias_compute_rsi`, arises due to the use of the `np.convolve` function with `mode='same'` when calculating the average gain and loss for the Relative Strength Index (RSI). Here's a detailed explanation:

### Understanding the RSI Calculation
The RSI is a momentum indicator that measures the speed and change of price movements. It is typically calculated over a lookback period (e.g., 14 days) using the following steps:
1. Compute price changes (`delta`).
2. Separate positive changes (`gain`) and negative changes (`loss`).
3. Calculate the average gain and average loss over the specified period.
4. Compute the Relative Strength (RS) as `avg_gain / avg_loss`.
5. Compute RSI as `100 - (100 / (1 + RS))`.

In a proper RSI calculation, the average gain and loss for each time point should only use data available up to that point (i.e., past and current data, not future data).

### Forward-Looking Bias in the Function
The function uses `np.convolve` with `mode='same'` to compute the moving averages of `gain` and `loss`. This introduces forward-looking bias because:
- The `np.convolve` function with `mode='same'` computes a centered moving average, meaning it uses both **past and future data** relative to each time point to calculate the average.
- For a given time point `t`, the average gain/loss includes data from `t` to `t + period/2` (future prices), which would not be available in a real-time trading or analysis scenario.

This is problematic because, in practice, future price data is unknown at the time of calculation. Using future data to compute the RSI results in an unrealistic and overly optimistic indicator that does not reflect what could have been computed at the time.

### Example of the Issue
Suppose you are calculating the RSI for day `t` with a 14-day period:
- The `np.convolve` function with `mode='same'` takes the `gain` values from `t - period/2` to `t + period/2` (approximately `t - 7` to `t + 7` for a 14-day period).
- This means the RSI at day `t` incorporates price changes from up to 7 days in the future (e.g., `t + 1` to `t + 7`), which are not available in a real-world setting.
- As a result, the RSI value is influenced by information that would not be accessible, leading to a forward-looking bias.

### Impact of Forward-Looking Bias
- **Backtesting**: If this RSI function is used in a trading strategy backtest, it will produce misleading results because it assumes knowledge of future price movements, potentially inflating performance metrics like returns or accuracy.
- **Real-Time Trading**: In live trading, this function cannot be used as-is because future data is unavailable, making the RSI values impractical to compute.

### How to Fix the Forward-Looking Bias
To eliminate the bias, the function should use a **causal** moving average that only considers past and current data. This can be achieved by:
1. Using a proper rolling average method that only looks backward.
2. Replacing `np.convolve` with a method like `pandas.DataFrame.rolling` or adjusting the convolution to avoid future data.

Here’s a corrected version of the function using `pandas`:

```python
import pandas as pd
import numpy as np

def compute_rsi(df, column="adjusted_close", period=14):
    delta = df[column].diff()
    gain = np.where(delta > 0, delta, 0)
    loss = np.where(delta < 0, -delta, 0)
    
    # Use rolling mean to compute averages using only past data
    avg_gain = pd.Series(gain).rolling(window=period, min_periods=period).mean()
    avg_loss = pd.Series(loss).rolling(window=period, min_periods=period).mean()
    
    rs = avg_gain / avg_loss
    rsi = 100 - (100 / (1 + rs))
    return rsi
```

### Key Improvements
- The `pandas.Series.rolling` method with `window=period` computes the average using only the data from `t - period + 1` to `t`, ensuring no future data is included.
- This makes the RSI calculation causal and suitable for both backtesting and real-time applications.

### Summary
The original function introduces forward-looking bias because `np.convolve` with `mode='same'` uses future price data to compute the RSI, which is unrealistic. This can lead to misleading results in backtesting or analysis. By switching to a backward-looking method like `pandas.Series.rolling`, the bias is eliminated, ensuring the RSI only uses data available at the time of calculation.
